src/js/components/settings/integrations.js

```diff
@@ -0,0 +1,142 @@
+
+  const updateConnectionConfig = ({ target: { value = '' } }) => setConnectionConfig(value);
+
+  return (
+    <>
+      <h3 className="margin-bottom-none">{EXTERNAL_PROVIDER[provider].title}</h3>
+      <div className="flexbox space-between relative" style={{ alignItems: 'flex-start', maxWidth }}>
+        <TextField
+          disabled={!isEditing}
+          InputLabelProps={{ shrink: !!connectionConfig }}
+          label={`${EXTERNAL_PROVIDER[provider].title} connection string`}
+          onChange={updateConnectionConfig}
+          value={connectionConfig}
+          style={{ minWidth: 500 }}
+        />
+        <div className="margin-left margin-top flexbox center-aligned">
+          {isEditing ? (
+            <>
+              <Button onClick={onCancelClick} style={{ marginRight: 10 }}>
+                Cancel
+              </Button>
+              <Button variant="contained" onClick={onSaveClick} disabled={connectionString === connectionConfig}>
+                Save
+              </Button>
+            </>
+          ) : (
+            <>
+              <Button onClick={onEditClick} style={{ marginRight: 10 }}>
+                Edit
+              </Button>
+              <Button onClick={onDeleteClick}>Delete</Button>
+            </>
+          )}
+        </div>
+        {isDeleting && (
+          <Confirm type="integrationRemoval" classes="confirmation-overlay" action={onDeleteConfirm} cancel={() => setIsDeleting(false)} style={{}} />
+        )}
+      </div>
+    </>
+  );
+};
+
+const emptyIntegration = {
+  connectionString: '',
+  provider: ''
+};
+
+export const Integrations = ({ integrations, changeIntegration, deleteIntegration }) => {
+  const [availableIntegrations, setAvailableIntegrations] = useState([]);
+  const [configuredIntegrations, setConfiguredIntegrations] = useState([]);
+  const [isCreating, setIsCreating] = useState(false);
+  const [currentValue, setCurrentValue] = useState('');
+
+  useEffect(() => {
+    setAvailableIntegrations(
+      Object.values(EXTERNAL_PROVIDER).reduce((accu, provider) => {
+        if (provider.enabled && !integrations.some(integration => integration.provider == provider.key)) {
+          accu.push(provider);
+        }
+        return accu;
+      }, [])
+    );
+    setConfiguredIntegrations(integrations);
+  }, [integrations]);
+
+  const onConfigureIntegration = ({ target: { value: provider } }) => {
+    setCurrentValue(provider);
+    setConfiguredIntegrations([{ ...emptyIntegration, provider }, ...configuredIntegrations]);
+    setIsCreating(true);
+  };
+
+  const onCancelClick = ({ connectionString }) => {
+    setIsCreating(!!connectionString);
+    setCurrentValue('');
+    setConfiguredIntegrations(integrations);
+  };
+
+  const onSaveClick = integration => {
+    changeIntegration(integration);
+    setCurrentValue('');
+    return setIsCreating(isCreating && !integration.connectionString ? false : isCreating);
```

Comment:
This is not working here. I use, as a test value, `HostName=test.azure-devices.net;SharedAccessKeyName=iothubowner;SharedAccessKey=ZHVtbXkgLW4K`.
This is what I get when I try to save the connection string:

![image](https://user-images.githubusercontent.com/1295287/142353908-11a05e46-e98c-4586-9e39-59e1571c7f05.png)

The connection string is correctly saved in the backend/database:

```
> db.settings.find().pretty()
{
	"_id" : ObjectId("6195dacc6c2c018095343290"),
	"connection_string" : {
		"access_key" : BinData(0,"ZHVtbXkgLW4K"),
		"hostname" : "test.azure-devices.net",
		"name" : "iothubowner"
	},
	"tenant_id" : ""
}
```

Despite it, though, refreshing still shows the "Add new integration" dropdown:

![image](https://user-images.githubusercontent.com/1295287/142354028-74743ce3-df34-4cd3-8f1b-9e79cee7c7fa.png)

Reply:
Thanks for that connection string - I tried some others I found, but none would validate - but using your format it should work now. However deleting the integration is not possible, as the validation rejects empty strings too 😕